fix_1165_unix_test_oddities_0.27

clanmills opened the pull request:
See #1165.

Unix Issues being caused by binary UserComment in exiv2-canon-powershot-s40.jpg and  exiv2-nikon-d70.jpg

clanmills committed:
Unix Issue is caused by binary UserComment in exiv2-canon-powershot-s40.jpg and exiv2-nikon-d70.jpg

clanmills committed:
Ensure $ make unit_test shows neither "Unsupported date format" nor "Unsupport time format".

clanmills committed:
Don't use diff option --text on UNIX in exiv2-test.sh

clanmills committed:
Fix exiv2-test.out

clanmills committed:
Fix preview-test.sh on UNIX.

clanmills committed:
exiv2-test.sh don't use sed -E option.

clanmills committed:
Restore data/preview-test.out

clanmills committed:
tiff-test.sh do not use sed -E

clanmills committed:
Restore make bash_tests on Unix == Linux.  test/Makefile merge TESTS1 and TEST2.  CMakeLists.txt remove alltests

clanmills committed:
Makefile polishing.

clanmills committed:
Fixing python_tests

clanmills committed:
Fixing tiff-test.sh to run on UNIX.  (cut, sed, grep have slightly different syntax on UNIX).

clanmills committed:
More Makefile polishing.

clanmills committed:
Fixing xmpparser-test.sh on Unix by disabling it (lots of diff/newline messages).

clanmills committed:
More Makefile polishing

clanmills committed:
bdiff only on SunOS

piponazo commented on tests/bugfixes/redmine/test_issue_443.py:
```diff
@@ -50,9 +48,8 @@ class CorruptedIopDir(metaclass=system_tests.CaseMeta):
 Exif.Thumbnail.YResolution                   Rational    1  72
 Exif.Thumbnail.ResolutionUnit                Short       1  inch
-Exif.Thumbnail.JPEGInterchangeFormat         Long        1  1628
+Exif.Thumbnail.JPEGInterchangeFormat         Long        1  1604
```
How is possible that the output of this test has changed ? According to the git history the input file for this test was removed or changed (`$data_path/exiv2-bug443.jpg`).  Was this intentional? I am a bit lost here.

piponazo reviewed (commented):
I am a bit lost in these changes. The description of #1165 only mentions that some tests are not running on some Unix systems, but I see here many changes in binary files (removal & addition of few jpeg files) and also in the expected output of tests running in every platform.

Could you please give some more information about what was changed here and why you needed to change the binary files? :innocent:

clanmills commented:
Sure.  The primary problem on UNIX is caused by "UserComment".  Many of our test files have user comments with binary stuff and that is reported differently on UNIX from Linux.  So, I've removed the "UserCommment" tag from some of our test files and that caused a ripple of trouble for the whole test suite.

The reason I became interested in doing Exiv2 v0.27.3 was because of #1046.  I had a discussion with Phil (@boardhead) and learnt new information about the specification of `UserComment`.  I believe the previous (0.27.2 and earlier) implementation of UserComment is suspect.  #1107 includes new tests and revised man page to explain how UserComment works.  It's a good thing to clean up our testing concerning UserComment.

The mechanism for UserComment is also used by GPSProcessingMethod and caused ripples in samples/geotag.cpp which were addressed in #1193.

The changes in the test suite code involve the utilities `sed`, `diff`, `cut` and `grep`.  Subtle differences.  SunOS has the command `bdiff`.

clanmills commented:
@piponazo May I ask you a question?  Why is codecov/project failing this build.  It's complaining about code in value.cpp which deals with UserComment!  (of course).  But what is he saying?

https://codecov.io/gh/Exiv2/exiv2/compare/338176b23943d7e6a772fff5c19f06d939a804e4...17e110cd5417503d473a301f9bc3c36c85ffc2b5

clanmills commented:
@piponazo I've been dreaming about UserComment.  For sure there's something wrong in the past.  On 'master' the reference output in test/data/exifdata-test.out has:

```
Exif.Photo.UserComment                        0x9286 Exif         Undefined 264                                                                                                                                                                                                                                                                 
```
There's lots of binary in the 264 bytes of the Undefined UserComment.

On UNIX, this is reported as:
```
Exif.Photo.UserComment                        0x9286 Exif         Undefined 264 
```
UNIX reports nothing from the 264 bytes.

Using my `dmpf` utility, the 264 Undefined bytes in the UserComment are all `nul nul ..... lf`.  The output is being performed by c_str().  Linux is outputting in binary.  On Unix, it terminates the string at the first nul.

```
  0x5dc0    24000: 286 Exif          ->  32 38 36 20 45 78 69 66 20 20 20 20 20 20 20 20
  0x5dd0    24016:  Undefined 264 .  ->  20 55 6e 64 65 66 69 6e 65 64 20 32 36 34 20 00
  0x5de0    24032: ................  ->  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x5df0    24048: ................  ->  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x5e00    24064: ................  ->  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x5e10    24080: ................  ->  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x5e20    24096: ................  ->  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x5e30    24112: ................  ->  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x5e40    24128: ................  ->  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x5e50    24144: ................  ->  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x5e60    24160: ................  ->  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x5e70    24176: ................  ->  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x5e80    24192: ................  ->  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x5e90    24208: ................  ->  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x5ea0    24224: ................  ->  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x5eb0    24240: ................  ->  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x5ec0    24256: ................  ->  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x5ed0    24272: ................  ->  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 0a
```

It should work like this:
```
2063 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ exiv2 -M'set Exif.Photo.UserComment 54321' ~/Stonehenge.jpg 
2064 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ exiv2 -g comment/i ~/Stonehenge.jpg 
Exif.Photo.UserComment                       Undefined  13  54321
2065 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ exiv2 -pR ~/Stonehenge.jpg | grep -i comment 
         428 | 0x9286 UserComment                  | UNDEFINED |       13 |       820 | ........54321
2066 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ 
```
There should be 8 bytes to define the charset, followed by the data of the string.

So what are all those `nul` bytes in the test file and why are they being reported in binary?

On Saturday evening, I thought "the UserComment metadata is suspect, I'll remove it from the reference files.".  Perhaps this is the wrong call.

I'll get to the bottom of this.  I always do.

piponazo commented:
Thanks for the explanation @clanmills. Honestly I did follow the details of the discussion you had in #1046, so it is nice that you wrote some documentation about it in #1107. 

Regarding you question about codecov, I normally do not pay too much attention to it unless the diff is very big. It is normal to have some small variations in the code coverage when refactoring code. And I also noticed that codecov sometimes does not take the right parent branch as reference to compute the diff.

clanmills commented:
@piponazo It's always nice to work with you.  You are very perceptive and clever.  And so is Dan.

I know how binary is getting into the output.  test/exifdata-test.sh uses a helper application samples/exifdata-test.cpp to perform various tricks.  And that program's `print()` function can pump binary into the output.  Why does it not output binary on UNIX?  Don't know yet.

More interestingly, it all has to do with UserComment.  There's special code in actions.cpp to format UserComment.  That doesn't feel right.  value.toString() should take care of that, so that applications (including actions.cpp, exifdata-test.cpp and exifprint.cpp) format comments correctly.  When I updated the code in actions.cpp to respect GPSAreaInformation, I put a negative comment in the code and the git/log.

I don't want to do maintenance on `bash_tests`, however I don't like the concept of samples/exifdata-test.cpp.  All testing should happen via samples/exiv2.cpp.  There are several test helpers documented in README-SAMPLES.md 

Once I am at the bottom of this, I'll probably restore the "old" test files, close this PR and generate a new PR.  I don't know the implication of closed PRs on the size of the git repository.  GitHub can restore deleted branches, so they are stored somewhere.  I hope they are not stored in every clone.

I've also been dreaming about codecov.  I think is says "You are no longer exercising this code.".  How useful to know.  The tests in #1107 should be using that.

As always, there are hidden puzzles to be hunted and fixed.  Today will be fun.  I'm enjoying this.

clanmills commented:
@piponazo I'm really delighted to let you know that I've done this a different way and solved several other puzzles at the same time.  I haven't modified any of the test images.  They are the same as master and all previous releases.

We should never output `metadatum.value()`.  That's wrong.  We should output `metadatum.toString()`.  And I've updated the commentValue::toString() to report the character encoding.  So I can throw away about 50 lines of strange code in actions.cpp concerning comments.

Tomorrow I will review `exiv2 --binary` in `actions.cpp` as it is strangely complicated.  It should be as simply as `exiv2 --hex`.

I have the revised code stored on my laptop and has been successfully built on all platforms.  I think we can close and delete this PR without making our repos bigger.  Can I open a new PR and submit this tomorrow?

piponazo commented:
It is great to read about your discoveries 👏 

Feel free to close the PR, it will not have an impact on the git repository size. What it takes space in the repository is the opened branches we might forget to close. I think that Github offers a button to delete the branch once you close a PR without merging it (I am not sure right now). But even if that is not the case, we can always remove the branches that are not merged in stable branches (master or 0.27-maintenance) later. From time to time, I check the branches I left behind to clean them up. I try to not touch branches from others, since I do not know if they would resume the work there at some point.

I am looking forward to see the changes you did! 😉

clanmills commented:
I'm closing this PR and deleting the branch `fix_1165_unix_test_oddities_0.27`

The branch `fix_1165_UNIX_test_0.27` uses a simpler and more flexible approach.
